File: Clojure/problem017.clj
#! /usr/bin/env clojure
; Problem 17 - Project Euler
; http://projecteuler.net/index.php?section=problems&id=017

(defn count-string [s]
  (-> s
      (clojure.string/split #" ")
      (clojure.string/join)
      (count)))

(def count-1-to-9
  (count-string "one two three four five six seven eight nine"))

(def count-10-to-19
  (count-string "ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen"))

(def count-20-to-99
  (-> "twenty thirty forty fifty sixty seventy eighty ninety"
      (clojure.string/split #" ")
      (as-> x
            (map #(+ (* (count %) 10) count-1-to-9) x)
            (reduce + x))))

(def count-1-to-99
  (+ count-1-to-9 count-10-to-19 count-20-to-99))

(def count-100-to-999
  (let [huns (count "hundred")
        ands (count "and")]
    (+ (* count-1-to-9 100)
       (* huns 900)
       (* ands (- 900 9))
       (* count-1-to-99 9))))

(def count-1000
  (count-string "one thousand"))

 (println (+ count-1-to-99
             count-100-to-999
             count-1000))

File: Go/problem017.go
package projecteuler

// #17 Number letter counts - Project Euler
// http://projecteuler.net/index.php?section=problems&id=17
import (
	"log"
	"strings"
)

func P017() {
	s1to9 := len(strings.Join([]string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}, ""))
	s10to19 := len(strings.Join([]string{"ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"}, ""))
	s20to99 := len(strings.Join([]string{"twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"}, ""))*10 + s1to9*8

	s1to99 := s1to9 + s10to19 + s20to99

	hundred := len("hundred")
	and := len("and")

	s100to999 := s1to9*100 + (hundred*100+and*99+s1to99)*9
	s1000 := len(strings.Join([]string{"one", "thousand"}, ""))

	ans := s1to99 + s100to999 + s1000

	log.Println(ans)
}
